### PR TITLE
Harmonics

### PR DESCRIPTION
**Pull request by bmcfee**
This PR implements a harmonic estimator.

This can be used as a subroutine to estimate pitch or tempo salience.

---
This change is [Reviewable](https://reviewable.io/reviews/librosa/librosa/405)

### CODE REVIEW

**Comment by bmcfee** [comment]
Note @stefan-balke @justinsalamon:

I think it might be a royal pain to get this working for instantaneous frequency measurements.  The complicating factor here is that the scipy interpolators are designed for fixed-grid measurement coordinates, so using the estimated bin centers from `ifgram` doesn't quite fit.

The only workaround I can think of is to estimate harmonics for each ifgram frame independently.  This will be absurdly slow.

**Comment by bmcfee** [comment]
Anyone up to CR this?

File: librosa/core/spectrum.py [hunk 1]
```diff
@@ -993,6 +993,123 @@ def fmt(y, t_min=0.5, n_fmt=None, kind='cubic', beta=0.5, over_sample=1, axis=-1
...
+    idx_out = [slice(None)] * harmonic_out.ndim
+
+    # Iterate over the harmonics range
+    for h_index, harmonic in enumerate(h_range):
+        idx_out[0] = h_index
+
+        # Iterate over frequencies
+        for f_index, frequency in enumerate(freqs):
+            # Offset the output axis by 1 to account for the harmonic index
+            idx_out[1 + axis] = f_index
```
Comment (bmcfee) [inline librosa/core/spectrum.py]: ooh, this is bad mojo if `axis=-1`.

**Comment by justinsalamon** [comment]
> I think it might be a royal pain to get this working for instantaneous frequency measurements.

IF measurements significantly improve pitch tracking accuracy, so without it the value of the method is limited. I can CR once I'm back in September (I'm being submitted to a strict no-work regime which I am breaking even by writing here) unless someone beats me to it.

**Comment by bmcfee** [comment]
Aaaaand it turns out that `interp2d` can't scale to anywhere close to the number of data points in a typical spectrogram.  Le sigh.

**Comment by bmcfee** [comment]
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by bmcfee** [comment]
I did a bit more poking around, and I think iterating over frames is the best (only) way to do harmonics for instantaneous frequencies.

That's a different enough use-case that I think it may be best handled in a separate function, and a separate PR; maybe #297?

**Comment by bmcfee** [comment]
Okay, refactored a bit and we now have a crude hack at time-varying frequency harmonic interpolation.  It's slower than molasses in january, and has occasional stability problems (due to lingering issues in frequency reassignment #137 , I suspect), but I think it's correct.  Still needs docs and tests.

**Comment by bmcfee** [comment]
Coverage is back up.  I think this is ready for CR, maybe @stefan-balke can take a look?

I looked into accelerating the frame-wise interpolation loop with numba, but it fails because numba can't handle numpy fancy (tuple of slice) indexing (yet).  We can revisit this in the future, I suppose.

**Comment by stefan-balke** [comment]
> Coverage is back up. I think this is ready for CR, maybe @stefan-balke can take a look?

Yup, will do so.

**Comment by stefan-balke** [comment]
Sorry for the delay here, ICASSP is eating my cycles. But will get to this ASAP.

**Comment by bmcfee** [comment]
@stefan-balke decompressed from icassp yet? :grimacing:

**Comment by stefan-balke** [comment]
Teaching at the Jazzomat Summer School in Weimar...

Next week it's all over and I will have time for this!

**Comment by stefan-balke** [comment]
*[librosa/core/harmonic.py, line 17 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpbpT0wP1_OpLGPR20:-KSpbpT0wP1_OpLGPR21:budp8a) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L17)):*
> ```Python
>     Parameters
>     ----------
>     X : np.ndarray, real-valued
> ```

small x

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by stefan-balke** [comment]
*[librosa/core/harmonic.py, line 41 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpc12I0qKN-yEuCqWf:-KSpc12I0qKN-yEuCqWg:bpa6kld) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L41)):*
> ```Python
>     Returns
>     -------
>     X_harm : np.ndarray, shape=(len(h_range), [X.shape])
> ```

harmonic_out

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by stefan-balke** [comment]
*[librosa/core/harmonic.py, line 150 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcDe0s3ZvtuiT23Ok:-KSpcDe0s3ZvtuiT23Ol:bgnl3o9) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L150)):*
> ```Python
>     axis : int
>         The axis along which to compute harmonics
> ```

Return value?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by stefan-balke** [comment]
*[librosa/core/harmonic.py, line 232 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcFncmALprT94hnTs:-KSpcFncmALprT94hnTt:bvllzux) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L232)):*
> ```Python
> 
> def harmonics_2d(harmonic_out, x, freqs, h_range, kind='linear', fill_value=0, axis=0):
> ```

missings docs

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by stefan-balke** [comment]
*[librosa/core/harmonic.py, line 248 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcOzmyfYFpagIIvn3:-KSpcOzmyfYFpagIIvn4:bobpoza) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L248)):*
> ```Python
>         harmonics_1d(harmonic_out[idx_out], x[idx_in], freqs[idx_freq],
>                      h_range, kind=kind, fill_value=fill_value,
>                      axis=axis)
> ```

Either both functions return something or none of them, or?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by stefan-balke** [comment]
Only first run on formal stuff. Next will be unit tests, functionality and running the examples.

**Comment by bmcfee** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[librosa/core/harmonic.py, line 41 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpc12I0qKN-yEuCqWf:-KSwcug-4TNFoqIdCKU1:b-8ii7bf) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L41)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary>
> harmonic_out
</details>
The internal variable name isn't important here; but it should be `x_out`

---

*[librosa/core/harmonic.py, line 150 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcDe0s3ZvtuiT23Ok:-KSwdAGF7CamsJn0YGT1:b-fimelf) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L150)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary>
> Return value?
</details>
Three is no return value; it operates in-place.

---

*[librosa/core/harmonic.py, line 248 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcOzmyfYFpagIIvn3:-KSweFL9mhXZYVWfTswE:b-l529t4) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L248)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary>
> Either both functions return something or none of them, or?
</details>
Neither of them do; they operate in-place.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by bmcfee** [comment]
@stefan-balke pinging back: i updated the docstrings.

I think there was some confusion in the specification here. Only `harmonics` is a user-facing function (appears in `__all__` and the docs).  The internal functions `harmonics_?d` are not meant for external use; that's why they operate in-place instead of returning a value.

**Comment by stefan-balke** [comment]
*[librosa/core/harmonic.py, line 248 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcOzmyfYFpagIIvn3:-KTUms3qh1EsnMyzIivN:b-lwrrwi) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L248)):*
<details><summary><i>Previously, bmcfee (Brian McFee) wrote…</i></summary>
> Neither of them do; they operate in-place.
</details>
Well, that's what I thought but you have one in this function. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by stefan-balke** [comment]
*[librosa/core/harmonic.py, line 229 at r7](https://reviewable.io:443/reviews/librosa/librosa/405#-KTUn4YiTWzEkwkFJaaB:-KTUn4YiTWzEkwkFJaaC:bi6uelb) ([raw file](https://github.com/librosa/librosa/blob/4c359c20ef558d5e117f7bb8d2ae8576dcfd67b2/librosa/core/harmonic.py#L229)):*
> Well, that's what I thought but you have one in this function. :)

that one should go then

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by bmcfee** [comment]
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[librosa/core/harmonic.py, line 248 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcOzmyfYFpagIIvn3:-KTUsly9vmElNpJT6Tiw:b-ypbwya) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L248)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary>
> Well, that's what I thought but you have one in this function. :)
</details>
Have one what?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by bmcfee** [comment]
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[librosa/core/harmonic.py, line 17 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpbpT0wP1_OpLGPR20:-KTV21vqwg_gHE-U1cX-:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L17)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary>
> small x
</details>
Done.

---

*[librosa/core/harmonic.py, line 41 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpc12I0qKN-yEuCqWf:-KTV23Co9iaAvngxy6v5:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L41)):*
<details><summary><i>Previously, bmcfee (Brian McFee) wrote…</i></summary>
> The internal variable name isn't important here; but it should be `x_out`
</details>
Done.

---

*[librosa/core/harmonic.py, line 150 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcDe0s3ZvtuiT23Ok:-KTV24OsOmx1KAVPEWXN:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L150)):*
<details><summary><i>Previously, bmcfee (Brian McFee) wrote…</i></summary>
> Three is no return value; it operates in-place.
</details>
Done.

---

*[librosa/core/harmonic.py, line 232 at r6](https://reviewable.io:443/reviews/librosa/librosa/405#-KSpcFncmALprT94hnTs:-KTV2A_Kq7R5LFOTDxTl:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/b12408a92abd07077150973c9f58de30600bd826/librosa/core/harmonic.py#L232)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary>
> missings docs
</details>
Done.

---

*[librosa/core/harmonic.py, line 229 at r7](https://reviewable.io:443/reviews/librosa/librosa/405#-KTUn4YiTWzEkwkFJaaB:-KTV27-CEHbwbkkLVaNe:b-hgxqbp) ([raw file](https://github.com/librosa/librosa/blob/4c359c20ef558d5e117f7bb8d2ae8576dcfd67b2/librosa/core/harmonic.py#L229)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary>
> > Well, that's what I thought but you have one in this function. :)
> 
> that one should go then
</details>
I don't understand this comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by stefan-balke** [comment]
*[librosa/core/harmonic.py, line 229 at r7](https://reviewable.io:443/reviews/librosa/librosa/405#-KTUn4YiTWzEkwkFJaaB:-KTY7d1w2JlKGVee289-:b9umcx6) ([raw file](https://github.com/librosa/librosa/blob/4c359c20ef558d5e117f7bb8d2ae8576dcfd67b2/librosa/core/harmonic.py#L229)):*
<details><summary><i>Previously, bmcfee (Brian McFee) wrote…</i></summary>
> I don't understand this comment.
</details>
Nevermind, I looked at the wrong commit somehow...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Comment by stefan-balke** [comment]
So, I finally got some time to look through the code, run the tests and the examples. All seems to work and the example plots look reasonable. The only thing I would suggest is, that the concept of the approach should be explained more clearly. Although this will rather serve as a backend method for higher-level functions, I still think the concept should be made clearer. #297 has lots of thoughts in it, maybe @bmcfee could boil it down to the core idea and add it to the docs of `harmonics()`? Or instead of having an abstract explanation, we add a little toy example.

output of the examples

**Comment by bmcfee** [comment]
Sure, I can update the docs and add more examples.

This PR was never meant to implement "salience" per se, just the backend as you mention.  #297 will remain open

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405#-:-KTZ__tI7PPo6xL1ltm5:b-poyppo)*

**Comment by bmcfee** [comment]
> Sure, I can update the docs and add more examples.

I updated the docs a little, but I'm not sure what else can be added to the examples to make them more clear.  Any suggestions?  Or should we call it a day and merge?

**Comment by stefan-balke** [comment]
> Or should we call it a day and merge?



I was temped to create a Trump-meme and call it Trump's day :)

**Comment by bmcfee** [comment]
Reviewed 1 of 4 files at r5, 1 of 1 files at r8, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/405)*

**Pull request merged**